### PR TITLE
[MISUV-9357] Add additional cessation date validation with contextual t…

### PR DESCRIPTION
**Pull request by Mbrookesy**
…ax year

### CODE REVIEW

**Comment by Mbrookesy** [comment]
retest this please

**Comment by Mbrookesy** [comment]
retest this please

File: app/forms/incomeSources/cease/CeaseIncomeSourceEndDateFormProvider.scala [hunk 1]
```diff
@@ -51,18 +51,18 @@ class CeaseIncomeSourceEndDateFormProvider extends Mappings {
 
     val minimumDate: Constraint[LocalDate] = incomeSourceType match {
       case UkProperty =>
-        val ukStartDate = user.incomeSources.properties.filter(_.isUkProperty).filter(!_.isCeased).flatMap(_.tradingStartDate)
-          .headOption.getOrElse(LocalDate.MIN)
+        val ukStartDate = user.incomeSources.properties.filter(_.isUkProperty).filter(!_.isCeased).map(_.getTradingStartDate).headOption.getOrElse(LocalDate.MIN)
```
Comment (michael-hmrc) [inline app/forms/incomeSources/cease/CeaseIncomeSourceEndDateFormProvider.scala:54]: heya just wondering you know they were .filtering twice here, do you think it would be worth combining it into a single filter? Or is it different when it's filtered twice?

File: app/models/incomeSourceDetails/BusinessDetailsModel.scala [hunk 2]
```diff
@@ -36,6 +37,13 @@ case class BusinessDetailsModel(incomeSourceId: String,
   def isCeased: Boolean = cessation.exists(_.date.nonEmpty)
 
   def isOngoingSoleTraderBusiness(id: String): Boolean = !isCeased && id == incomeSourceId
+
+  def getTradingStartDate: LocalDate = (tradingStartDate, contextualTaxYear) match {
+    case (Some(startDate), None) => startDate
+    case (None, Some(taxYear)) => LocalDate.of(taxYear - 1, 4, 6)
+    case (Some(startDate), Some(_)) => startDate
+    case (None, None) => LocalDate.MIN
+  }
```
Comment (michael-hmrc) [inline app/models/incomeSourceDetails/BusinessDetailsModel.scala:46]: not 100% sure but thinking `getTradingStartDate` can be simplified?
```
def getTradingStartDate: LocalDate = (tradingStartDate, contextualTaxYear) match {
    case (None, None) => LocalDate.MIN
    case (None, Some(taxYear)) => LocalDate.of(taxYear - 1, 4, 6)
    case _ => startDate
  }
```

File: app/forms/incomeSources/cease/CeaseIncomeSourceEndDateFormProvider.scala (hunk 1 above)
Comment (Mbrookesy) [reply app/forms/incomeSources/cease/CeaseIncomeSourceEndDateFormProvider.scala:54]: Honestly i'm not sure, I can try to see if it'll break anything and get back to you on that

File: app/models/incomeSourceDetails/BusinessDetailsModel.scala (hunk 2 above)
Comment (avelx) [reply app/models/incomeSourceDetails/BusinessDetailsModel.scala:46]: calculation of tax year end date is redundant as its already part of taxYear type


```
  def toFinancialYearStart: LocalDate = {
    LocalDate.of(this.startYear, April, Sixth)
  }

  def toFinancialYearEnd: LocalDate = {
    LocalDate.of(this.endYear, April, Fifth)
  }
```

Comment (Mbrookesy) [reply app/models/incomeSourceDetails/BusinessDetailsModel.scala:46]: We aren't using the TaxYear class though so I'm not sure if that's relevant here?

Comment (avelx) [reply app/models/incomeSourceDetails/BusinessDetailsModel.scala:46]: I would think that next code would be bit easy to ready:

`TaxYear(... ).toFinancialYearStart `

File: app/forms/incomeSources/cease/CeaseIncomeSourceEndDateFormProvider.scala (hunk 1 above)
Comment (Mbrookesy) [reply app/forms/incomeSources/cease/CeaseIncomeSourceEndDateFormProvider.scala:54]: yeah it looks like it breaks it because it needs to filter through to one type of income source first before it checks inside for what is ceased

**Comment by Mbrookesy** [comment]
retest this please

**Review by michael-hmrc — DISMISSED** [review]

File: test/controllers/manageBusinesses/cease/IncomeSourceEndDateControllerSpec.scala [hunk 3]
```diff
@@ -213,6 +213,80 @@ class IncomeSourceEndDateControllerSpec extends MockAuthActions with MockSession
                 redirectLocation(result) shouldBe Some(redirectLocationResult)
               }
             }
+            "display before tax year error" when {
+              "form is completed successfully" in {
+                setupMockSuccess(mtdRole)
+                enable(IncomeSourcesFs)
```
Comment (chandraAndem) [inline test/controllers/manageBusinesses/cease/IncomeSourceEndDateControllerSpec.scala:219]: this should be IncomeSourceNewFs i.e., new journey but this issue is existing from before and I'll fix them all in my PR.

File: test/controllers/manageBusinesses/cease/IncomeSourceEndDateControllerSpec.scala [hunk 4]
```diff
@@ -213,6 +213,80 @@ class IncomeSourceEndDateControllerSpec extends MockAuthActions with MockSession
...
+                mockBothPropertyBothBusiness()
+                setupMockCreateSession(true)
+                if (incomeSourceType == SelfEmployment) {
+                  setupMockSetMultipleMongoData(Right(true))
+                } else {
+                  setupMockSetSessionKeyMongo(Right(true))
+                }
+                setupMockGetMongo(Right(Some(emptyUIJourneySessionData(IncomeSourceJourneyType(Cease, incomeSourceType)))))
+                val result: Future[Result] = action(fakeRequest.withFormUrlEncodedBody("value.day" -> "27", "value.month" -> "8",
+                  "value.year" -> "2025"))
```
Comment (chandraAndem) [inline test/controllers/manageBusinesses/cease/IncomeSourceEndDateControllerSpec.scala:266]: this will fail in after 27th Aug 2025, so could you change year to currentYear + 1 like we discussed earlier so that it will always be future date.

File: test/controllers/incomeSources/cease/IncomeSourceEndDateControllerSpec.scala [hunk 5]
```diff
@@ -312,6 +312,80 @@ class IncomeSourceEndDateControllerSpec extends MockAuthActions with MockSession
...
+              enable(IncomeSourcesFs)
+              mockBothPropertyBothBusiness()
+              setupMockCreateSession(true)
+              if (incomeSourceType == SelfEmployment) {
+                setupMockSetMultipleMongoData(Right(true))
+              } else {
+                setupMockSetSessionKeyMongo(Right(true))
+              }
+              setupMockGetMongo(Right(Some(emptyUIJourneySessionData(IncomeSourceJourneyType(Cease, incomeSourceType)))))
+              val result: Future[Result] = action(fakeRequest.withFormUrlEncodedBody("value.day" -> "27", "value.month" -> "8",
```
Comment (chandraAndem) [inline test/controllers/incomeSources/cease/IncomeSourceEndDateControllerSpec.scala:364]: this will fail in after 27th Aug 2025, so could you change year to currentYear + 1 like we discussed earlier so that it will always be future date.

**Review by chandraAndem — CHANGES_REQUESTED** [review]
Minor comments and also test related to minimum date is missing i.e., before 06-Apr-2015

**Comment by chandraAndem** [comment]
retest this please

**Comment by chandraAndem** [comment]
test this please

**Comment by chandraAndem** [comment]
test this please

**Review by jmountain-hmrc — APPROVED** [review]
